contracts/meta-oracles/DriftlessHistoricalPriceFeed.sol

```diff
@@ -0,0 +1,277 @@
+    }
+
+    /*
+     * Query linked list for specified days of data. Will revert if number of days
+     * passed exceeds amount of days collected. Will revert if not enough days of
+     * data logged.
+     *
+     * @param  _dataDays       Number of days of data being queried
+     * @returns                Array of historical price data of length _dataDays                   
+     */
+    function read(
+        uint256 _dataDays
+    )
+        external
+        view
+        returns (uint256[] memory)
+    {
+        return readList(
+            historicalPriceData,
+            _dataDays
+        );
+    }
+
+    /*
+     * Change medianizer in case current one fails or is deprecated. Only contract
+     * owner is allowed to change.
+     *
+     * @param  _newMedianizerAddress       Address of new medianizer to pull data from
+     */
+    function changeMedianizer(
+        address _newMedianizerAddress
+    )
+        external
+        onlyOwner
+    {
+        medianizerInstance = IMedian(_newMedianizerAddress);
+    }
+
+
+    /* ============ Private ============ */
+
+    /*
+     * Create initialValues array from _seededValues and the current medianizer price.
+     * Added to historicalPriceData in constructor.
+     *
+     * @param  _seededValues        Array of previous days' historical price values to seed
+     * @returns                     Array of initial values to add to historicalPriceData                  
+     */
+    function createInitialValues(
+        uint256[] memory _seededValues
+    )
+        private
+        returns (uint256[] memory)
+    {
+        // Get current value from medianizer
+        uint256 currentValue = uint256(medianizerInstance.read());
+
+        // Instantiate outputArray
+        uint256 seededValuesLength = _seededValues.length;
+        uint256[] memory outputArray = new uint256[](seededValuesLength.add(1));
+
+        // Take values from _seededValues array and add to outputArray
+        for (uint256 i = 0; i < _seededValues.length; i++) {
+            outputArray[i] = _seededValues[i];
+        }
+
+        // Add currentValue to outputArray
+        outputArray[seededValuesLength] = currentValue;
+
+        return outputArray;
+    }
+
+    /*
+     * Determine price to update LinkedList with. If within time tolerance then we take raw value from oracle,
+     * otherwise linearize current oracle price with last logged price to attempt to reduce potential error.
+     *
+     * @returns                     Price to update LinkedList with                  
+     */
+    function determineUpdatePrice()
+        private
+        returns (uint256)
+    {
+        // Get current medianizer value
+        uint256 medianizerValue = uint256(medianizerInstance.read());
+
+        // Since the nextAvailableUpdate timestamp was already update to prevent reentrancy we must subtract
+        // the updateFrequency then add the updateTolerance to get the timestamp after which we linearize
```

Comment:
This seems funky. Might be worth adding `nonReEntrant` and then updating the `nextAvailableUpdate` after-the-fact.